### ISSUE TITLE
Tokens leaking into logfiles -- additional locations

### ISSUE THREAD

**Issue by bbockelm** [xrootd/xrootd#2296]
This is a follow-up to #2222 - copying some tokens-in-logs instance I had noted in that ticket which didn't get fixed (didn't want to re-open a ticket marked as solved in release notes).

Here are the examples:

```
$ XRD_PELICANCERTFILE=~/.config/pelican/certificates/tlsca.pem XRD_PLUGINCONFDIR=~/projects/xrdcl-curl/build/release_dir/etc/xrootd/client.plugins.d/ ~/projects/xrootd/build/release_dir/bin/xrootd -c ~/.config/xrootd/xrootd-bbockelm-cache.cfg 2>&1 | grep foo1234
240516 17:14:56 5379 http_Protocol:  rc:24 got hdr line: Authorization: foo1234
240516 17:14:56 5379 unknown.1:31@localhost http_Req: Appended header fields to opaque info: 'authz=foo1234&pelican.timeout=3s'
[2024-05-16 17:14:56.807137 +0200][Debug  ][ExDbgMsg          ] [p0@eduroam-0197.desy.de:1095] MsgHandler created: 0x52e099c0 (message: kXR_stat (path: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, flags: none) ).
[2024-05-16 17:14:56.811735 +0200][Error  ][XRootD            ] [p0@eduroam-0197.desy.de:1095] Impossible to send message kXR_stat (path: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, flags: none). Trying to recover.
[2024-05-16 17:14:56.811746 +0200][Debug  ][XRootD            ] [p0@eduroam-0197.desy.de:1095] Handling error while processing kXR_stat (path: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, flags: none): [FATAL] Login failed.
[2024-05-16 17:14:56.811755 +0200][Debug  ][ExDbgMsg          ] [p0@eduroam-0197.desy.de:1095] Passing to the thread-pool MsgHandler: 0x52e099c0 (message: kXR_stat (path: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, flags: none) ).
[2024-05-16 17:14:56.811777 +0200][Debug  ][ExDbgMsg          ] [p0@eduroam-0197.desy.de:1095] Calling MsgHandler: 0x52e099c0 (message: kXR_stat (path: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, flags: none) ) with status: [FATAL] Login failed.
240516 17:14:56 5379 unknown.1:31@localhost Xrootd_Protocol: 0000 open rt /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s
240516 17:14:56 5379 XrdPfc_Cache: info Attach() root://u31@eduroam-0197.desy.de:1095//tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s
[2024-05-16 17:14:56.812103 +0200][Debug  ][File              ] [0x5400a1a8@root://u31@eduroam-0197.desy.de:1095//tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s&xrdcl.requuid=52BB5289-D4F8-420B-9EE9-5CB90FB38388] Sending an open command
[2024-05-16 17:14:56.812156 +0200][Debug  ][ExDbgMsg          ] [u31@eduroam-0197.desy.de:1095] MsgHandler created: 0x5400a320 (message: kXR_open (file: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, mode: 00, flags: kXR_open_read kXR_async kXR_retstat ) ).
[2024-05-16 17:14:56.817051 +0200][Error  ][XRootD            ] [u31@eduroam-0197.desy.de:1095] Impossible to send message kXR_open (file: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, mode: 00, flags: kXR_open_read kXR_async kXR_retstat ). Trying to recover.
[2024-05-16 17:14:56.817059 +0200][Debug  ][XRootD            ] [u31@eduroam-0197.desy.de:1095] Handling error while processing kXR_open (file: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, mode: 00, flags: kXR_open_read kXR_async kXR_retstat ): [FATAL] Login failed.
[2024-05-16 17:14:56.817066 +0200][Debug  ][ExDbgMsg          ] [u31@eduroam-0197.desy.de:1095] Passing to the thread-pool MsgHandler: 0x5400a320 (message: kXR_open (file: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, mode: 00, flags: kXR_open_read kXR_async kXR_retstat ) ).
[2024-05-16 17:14:56.817083 +0200][Debug  ][ExDbgMsg          ] [u31@eduroam-0197.desy.de:1095] Calling MsgHandler: 0x5400a320 (message: kXR_open (file: /tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s, mode: 00, flags: kXR_open_read kXR_async kXR_retstat ) ) with status: [FATAL] Login failed.
[2024-05-16 17:14:56.817150 +0200][Debug  ][File              ] [0x5400a1a8@root://u31@eduroam-0197.desy.de:1095//tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s&xrdcl.requuid=52BB5289-D4F8-420B-9EE9-5CB90FB38388] Open has returned with status [FATAL] Login failed: Got invalid login response.
[2024-05-16 17:14:56.817172 +0200][Debug  ][File              ] [0x5400a1a8@root://u31@eduroam-0197.desy.de:1095//tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s&xrdcl.requuid=52BB5289-D4F8-420B-9EE9-5CB90FB38388] Error while opening at u31@eduroam-0197.desy.de:1095: [FATAL] Login failed: Got invalid login response.
240516 17:14:56 5379 XrdPfc_IO: info initCachedStat successfully read size 0 and creation time 1715872496 from info file root://u31@eduroam-0197.desy.de:1095//tmp/xrootd/public/foo?authz=foo1234&pelican.timeout=3s
```

Note these other cases are from XrdCl and XrdPfc modules.  So, perhaps a quick review of those modules with the same obfuscation code.

**Comment by ccaffy** [comment]
Hi @bbockelm ,

I'm surprised, about the `http_Protocol` and `http_Req` lines that you provided. I have them obfuscated on my side:

```
240716 08:51:22 148346 http_Protocol:  rc:36 got hdr line: PUT /tmp/BIGFILE_5GB_dest HTTP/1.1
240716 08:51:22 148346 http_Protocol:  Parsing first line: PUT /tmp/BIGFILE_5GB_dest HTTP/1.1
240716 08:51:22 148346 http_Protocol:  rc:25 got hdr line: User-Agent: curl/7.76.1
240716 08:51:22 148346 http_Protocol:  rc:13 got hdr line: Accept: */*
240716 08:51:22 148346 http_Protocol:  rc:28 got hdr line: authorization:REDACTED
240716 08:51:22 148346 http_Protocol:  rc:12 got hdr line: SciTag: 65
240716 08:51:22 148346 http_Protocol:  rc:25 got hdr line: Content-Length: 5242880
240716 08:51:22 148346 http_Protocol:  rc:22 got hdr line: Expect: 100-continue
240716 08:51:22 148346 http_Protocol:  rc:2 got hdr line: 
```

Same for the `header2cgi` functionality:

```
240716 08:52:56 148407 unknown.1:27@xrootd http_Req: Appended header fields to opaque info: 'authz=REDACTED&scitag.flow=65'
240716 08:52:56 148407 unknown.1:27@xrootd Xrootd_Protocol: 0000 Bridge req=3010 dlen=57 blen=57
```

Which XRootD version are you running? I've tried this from the devel branch but also from the v5.7.0 tag ;)

Thank you.

Cheers,
Cedric

**Comment by abh3** [comment]
I think the particular complaints here are not from the server (well at least from the HTTP side).  What I see are correct complaints that the xroot client is spilling authz information into the log as it's used by Xcache. I also know that the Xcache plugin is doing the same itself in certain contexts. All of this is correctable but very tedious. It is even more depressing because it only happens when the log level is turned up for debugging which is not something that happens often, but we have to spend time addressing it.
________________________________

**Comment by ccaffy** [comment]
Still, Brian should not see any tokens leaking to the logs via those HTTP lines.

However, the fact that other tokens are leaking is "normal" because I did not obfuscate anything else than things related to HTTP ;)

**Comment by ccaffy** [comment]
Thanks @bbockelm now it's clear, I did not pay attention to the dates at all ;)

**Comment by amadio** [comment]
With the performance fix in #2314 we can now close this and add it to the next release. Thanks, @ccaffy!
